Remotion/Web/Core/Context/HttpContextStorageProviderNetCore.cs

```diff
@@ -0,0 +1,208 @@
+// License, or (at your option) any later version.
+// 
+// re-motion is distributed in the hope that it will be useful, 
+// but WITHOUT ANY WARRANTY; without even the implied warranty of 
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using System.Collections;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Web;
+using Remotion.Context;
+using Remotion.ServiceLocation;
+using Remotion.Utilities;
+
+namespace Remotion.Web.Context
+{
+#if !NETFRAMEWORK
+  [ImplementationFor(typeof(ISafeContextStorageProvider), Position = 0)]
+  public class HttpContextStorageProvider : ISafeContextStorageProvider, ISafeContextBoundaryStrategy
+  {
+    private struct HashTableAccessor
+    {
+      private class WrappedHashtable
+      {
+        public HttpContext Context { get; }
+        public Hashtable Data { get; }
+
+        public WrappedHashtable (HttpContext context, Hashtable data)
+        {
+          Context = context;
+          Data = data;
+        }
+      }
+
+      private const string c_dataDictionaryKey = "Rm.DD";
+
+      private readonly ThreadLocal<WeakReference<WrappedHashtable?>> _dataTable;
+
+      public HashTableAccessor (bool _)
+      {
+        _dataTable = new(() => new WeakReference<WrappedHashtable?>(null));
+      }
+
+      public bool TryGetCachedHashtable (HttpContext httpContext, [MaybeNullWhen(false)] out Hashtable hashtable)
+      {
+        if (TryGetCachedWrappedHashtable(httpContext, out var cachedWrappedHashtable))
+        {
+          hashtable = cachedWrappedHashtable.Data;
+          return true;
+        }
+
+        hashtable = null;
+        return false;
+      }
+
+      public Hashtable GetOrCreateHashtable (HttpContext context)
+      {
+        if (TryGetCachedHashtable(context, out var hashtable))
+        {
+          return hashtable;
+        }
+
+        return SetHashtable(context, new Hashtable());
+      }
+
+      public Hashtable? RemoveHashtable (HttpContext context)
+      {
+        if (TryGetCachedWrappedHashtable(context, out var cachedHashtable))
+        {
+          context.Items.Remove(c_dataDictionaryKey);
+          _dataTable.Value!.SetTarget(null!);
+        }
+
+        return cachedHashtable?.Data;
+      }
+
+      public Hashtable SetHashtable (HttpContext context, Hashtable hashtable)
+      {
+        var wrappedHashtable = new WrappedHashtable(context, hashtable);
+        context.Items[c_dataDictionaryKey] = wrappedHashtable;
+        _dataTable.Value!.SetTarget(wrappedHashtable);
+        return hashtable;
+      }
+
+      private bool TryGetCachedWrappedHashtable (HttpContext httpContext, [MaybeNullWhen(false)] out WrappedHashtable hashtable)
```

Comment:
If think we can simplify this method. We return a `WrappedHashtable`, but the usages always use only the `Data`. But if we change the out parameter to `Hashtable` it is effectively our public method above. Couldn't we just move the code into the public method and use that in the `RemoveHashtable` method? What do you think?